### DISCUSSION TITLE
populateWhere keeps return the filtered records from first request

### DISCUSSION THREAD

**Discussion by zgid123** [mikro-orm/mikro-orm#3640] [Q&A] [answered]
**Describe the bug**

I have 3 models: Foo, Bar and FoosBar. Foo has manyToMany association with Bar through FoosBar model.

```ts
// Foo.ts
@Entity()
export class Foo {
  @OneToMany(() => FoosBar, (foosBar) => foosBar.foo)
  foosBars = new Collection<FoosBar>(this);
}

// Bar.ts
@Entity()
export class Bar {
  @OneToMany(() => FoosBar, (foosBar) => foosBar.bar)
  foosBars = new Collection<FoosBar>(this);
}

// FoosBar.ts
export class FoosBar {
  @ManyToOne(() => Foo)
  foo: Foo;

  @ManyToOne(() => Bar)
  bar: Bar;
}
```

when get `foo` and eager loading its `foosBars` and filter `foosBars` with specific `bar`, the first time we get correct records, but the second time is like loading it from cache.

I use mikro-orm with NestJS. So the example code is like this

```ts
interface IQueryProps {
  barSlug: string;
}

@Controller()
class FoosController {
constructor(
  @InjectRepository(Foo)
    private readonly fooRepository: EntityRepository<Foo>,
  ) {}

  public getAll(@Query() query: IQueryProps): Promise<Foo[]> {
    const { barSlug } = query;

    return this.fooRepository.find({
      foosBars: {
        bar: {
          slug: barSlug,
        },
      },
    },
    {
      populate: ['foosBars', 'foosBars.bar'],
      populateWhere: PopulateHint.INFER,
      cache: false, // I assume it is loaded from cache, so I set this by false, but it seems to be not like that
    });
  }
}
```

**Expected behavior**

The result of `foosbars` records must be correct if the `barSlug` is changed, currently it will be like the first request.

**Additional context**
No

**Versions**

| Dependency | Version |
| - | - |
| node | 17.7.2 |
| typescript | 4.8.4 |
| mikro-orm | 5.4.2 |
| your-driver | postgresql |
| nestjs | 9.1.5 |

**Comment by B4nan** [comment, marked as answer]
>       cache: false, // I assume it is loaded from cache, so I set this by false, but it seems to be not like that

Result cache is disabled by default, this is a no-op. 

This sounds like you have miconfigured request context. How does your ORM config look like?

Also, to verify that, try adding `refresh: true` to the find options.

**Reply by zgid123** [reply]
my config is

```ts
{
    type,
    dbName: process.env.DB_NAME,
    allowGlobalContext,
    host: process.env.DB_HOST,
    user: process.env.DB_USER,
    entities: ['./dist/db/models'],
    entitiesTs: ['./src/db/models'],
    password: process.env.DB_PASSWORD,
    namingStrategy: MikroNamingStrategy,
    port: parseInt(process.env.DB_PORT || '', 10),
    debug: process.env.NODE_ENV === 'development',
    seeder: {
      glob: '!(*.d).{js,ts}',
      path: './dist/db/seeds',
      pathTs: './src/db/seeds',
    },
    migrations: {
      path: './dist/db/migrations',
      pathTs: './src/db/migrations',
    },
}
```

for the `MikroNamingStrategy`

```ts
import { plural } from 'pluralize';
import { UnderscoreNamingStrategy } from '@mikro-orm/core';

export class MikroNamingStrategy extends UnderscoreNamingStrategy {
  classToTableName(entityName: string, isPlural = true): string {
    return this.toTableName(entityName, isPlural);
  }

  joinKeyColumnName(entityName: string, referencedColumnName?: string): string {
    return `${this.classToTableName(entityName, false)}_${
      referencedColumnName || this.referenceColumnName()
    }`;
  }

  protected toTableName(entityName: string, isPlural = true): string {
    const tableName = entityName
      .replace(/([a-z])([A-Z])/g, '$1_$2')
      .toLowerCase();

    if (!isPlural) {
      return tableName;
    }

    return plural(tableName);
  }
}
```

**Reply by B4nan** [reply]
> allowGlobalContext

So you disable the validation and work with global context, right? What is happening is exactly the reason why you should not do that.

**Reply by zgid123** [reply]
actually, I don't understand that error

```
Using global EntityManager instance methods for context specific actions is disallowed. If you need to work with the global instance's identity map, use `allowGlobalContext` configuration option or `fork()` instead
```

My case is I have multiple NestJS apps and manage them using monorepo. I create a package lib to make a default config for MikroOrm for NestJS. And `MikroOrm` module keeps raising the error, so I add it for temporary solution

**Reply by zgid123** [reply]
the refresh option seems to work perfectly. Thanks for your help. I will dig this library later to understand all those options. But for now, I will use this as temporary solution to make the project run first.

Thanks for your help again.

**Reply by B4nan** [reply]
https://stackoverflow.com/questions/71117269/validation-error-using-global-entity-manager-instance-methods-for-context-speci/72799993#72799993

**Reply by zgid123** [reply]
After reading your post, I think I know the reason.

`Middlewares are executed only for regular HTTP request handlers`, I am using for gRPC. I think this is the main reason I need the option. I will use NestJS middleware to fix this issue.

Thanks again for helping me to understand this error.
